datafusion/substrait/src/logical_plan/consumer.rs

```diff
@@ -984,30 +984,64 @@ pub async fn from_substrait_rel(
 /// 1. All fields present in the Substrait schema are present in the DataFusion schema. The
 ///    DataFusion schema may have MORE fields, but not the other way around.
 /// 2. All fields are compatible. See [`ensure_field_compatability`] for details
-///
-/// This function returns a DataFrame with fields adjusted if necessary in the event that the
-/// Substrait schema is a subset of the DataFusion schema.
 fn ensure_schema_compatability(
-    table: DataFrame,
+    table_schema: DFSchema,
     substrait_schema: DFSchema,
-) -> Result<DataFrame> {
-    let df_schema = table.schema().to_owned().strip_qualifiers();
-    if df_schema.logically_equivalent_names_and_types(&substrait_schema) {
-        return Ok(table);
-    }
-    let selected_columns = substrait_schema
+) -> Result<()> {
+    substrait_schema
         .strip_qualifiers()
         .fields()
         .iter()
         .map(|substrait_field| {
             let df_field =
-                df_schema.field_with_unqualified_name(substrait_field.name())?;
-            ensure_field_compatability(df_field, substrait_field)?;
-            Ok(col(format!("\"{}\"", df_field.name())))
+                table_schema.field_with_unqualified_name(substrait_field.name())?;
+            ensure_field_compatability(df_field, substrait_field)
         })
-        .collect::<Result<_>>()?;
+        .collect::<Result<()>>()?;
+
+    Ok(())
+}
+
+/// This function returns a DataFrame with fields adjusted if necessary in the event that the
+/// Substrait schema is a subset of the DataFusion schema.
+fn apply_projection(table: DataFrame, substrait_schema: DFSchema) -> Result<LogicalPlan> {
```

Comment:
I'm finding the code flow here a little tricky to follow with regards to how we would handle the ReadRel::filter.

1. This function is consuming a `substrait_schema` that has already had the mask expressions applied to it.
2. This function creates the DataFusion TableScan from the `table`, which is what we would need to attach the `ReadRel::filter` to using something like `scan.filters = <converted filters>`
3. But to build the `<converted filters>`, we'll need access to the unmasked `substrait_schema` in order to correctly resolve any Substrait column references. 

It might be easier to follow along with this, and make it easier to re-use, if we changed this function to something like:
```rust
/// Given the table produced by ReadRel::read_type, apply the remaining ReadRel fields
fn convert_read_rel(table: DataFrame, read_rel: &ReadRel) -> Result<LogicalPlan> {
    let named_struct = read.base_schema.as_ref().ok_or_else(|| {
        substrait_datafusion_err!("No base schema provided for ReadRel")
    })?;
    let substrait_schema =
        from_substrait_named_struct(named_struct, extensions)?
            .replace_qualifier(table_reference);

    ensure_schema_compatability(
        table.schema().to_owned(),
        substrait_schema.clone(),
    )?;

    let t = table.into_unoptimized_plan();

    // Apply filter (for later)
    // Apply best-effort filter (for later)
    // Apply masking (already done)
    // Apply RelCommon::emit (for later, and yes a ReadRel can have both masking AND a emit remapping)

   todo!()
}
```

Effectively the responsibility of the caller is to produce the underlying table, and then this function applies the rest of the ReadRel properties which are common to all of the different read_types.

All that being said, this might be more scope than you intended to take and might be better as a follow-up. I'll leave it up to you.

Reply:
> this might be more scope than you intended to take and might be better as a follow-up

I think I crossed that line a long ago, it's fine :laughing: 

Can you clarify how is this different from the current flow (aside from it being refactored into another function)? My thought was also to eventually add it [here](https://github.com/apache/datafusion/blob/e8b2b2a695e657034824200f9f2403c01179a3fd/datafusion/substrait/src/logical_plan/consumer.rs#L769) before calling `apply_masking` on the schema. We can always obtain a TableScan from a df and mutate it. then `apply_projection` would be changed to accept TableScan directly instead of a DataFrame.

Or maybe I'm misunderstanding your point. Are you challenging the fact that masking is applied to the schema first, which is then applied to a TableScan? The rationale there was that base_schema and masking are both needed in conjunction to come up with column indices and if I did in two different steps, I'd have to remap already added projections which seemed more error-prone.

Reply:
Oh! You can attach a filter directly to the DataFrame? TIL. That addresses my data dependency concern, I thought you would only be able to apply the filter to the TableScan.